File: pointers/linkedlists/linkedlist_has_cycle.py
# LinkedList has cycle?

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        # attempt 2: (source: neetcode)
        # if there is no cycle, guaranteed to terminate 
        slow, fast = head, head
        while (fast != None and fast.next !=None):
            slow = slow.next
            fast = fast.next.next
            if (slow == fast):
                return True
        return False
        
        # attempt 1:
        # Below is an implementation that is slower
        # helper = set()
        # pointer = head
        # while (pointer!=None and pointer.next != None):
        #     if (pointer in helper):
        #         return True
        #     pointercopy = pointer
        #     helper.add(pointercopy)
        #     pointer = pointer.next
        # return False
    

File: pointers/linkedlists/linkedListHasCycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    let p1 = head;
    let p2 = head;
    while (p2 != null && p2.next !=null) {
        // thankfully has checks for object comparison
        // Note: this implementation of two pointers 
        // ... of speed 1 and 2 works because gcd(1,2) = 1
        p1 = p1.next;
        p2 = p2.next.next;
        if (p1 === p2) { 
            return true;
        }
    }
    return false; 
    // Initial implementation version::
    // below is a version that is slower::
    // let helper = new Set();
    // let pointer = head;
    // while (pointer!=null && pointer.next!=null) {
    //     // thankfully has checks for object comparison
    //     if (helper.has(pointer)) {
    //         return true;
    //     }
    //     helper.add(pointer);
    //     pointer = pointer.next;
    // }
    // return false; 
};
